posthog/hogql/printer.py

```diff
@@ -555,7 +592,132 @@ def visit_lambda(self, node: ast.Lambda):
     def visit_order_expr(self, node: ast.OrderExpr):
         return f"{self.visit(node.expr)} {node.order}"
 
+    def __get_optimized_property_group_compare_operation(self, node: ast.CompareOperation) -> str | None:
+        """
+        Returns a printed expression corresponding to the provided compare operation, if one of the operands is part of
+        a property group value and: the comparison can be rewritten so that it can be eligible for use by one or more
+        the property group's bloom filter data skipping indices, or the expression can be optimized to avoid reading the
+        property group's map ``values`` subcolumn when doing comparisons to NULL values.
+        """
+        if not self.context.modifiers.usePropertyGroups == UsePropertyGroups.OPTIMIZED:
+            return None
+
+        def resolve_field_type(expr: ast.Expr) -> ast.Type | None:
+            expr_type = expr.type
+            while isinstance(expr_type, ast.FieldAliasType):
+                expr_type = expr_type.type
+            return expr_type
+
+        if node.op in (ast.CompareOperationOp.Eq, ast.CompareOperationOp.NotEq):
+            # For commutative operations, we can rewrite the expression with parameters in either order without
+            # affecting the result.
+            # NOTE: For now, this only works with comparisons to constant values directly, but this could be improved so
+            # that any expression that doesn't reference a field as part of the expression (and therefore can be
+            # resolved to a constant value during the initial stages of query execution, e.g. ``lower(concat('X','Y'))``
+            # and similar) is supported to improve the usefulness of the index -- but we'll need to take care to handle
+            # expressions that return NULL values or empty strings. (The same applies to ``In`` comparisons below, too.)
```

Comment:
This is tricky to actually do since we would have to handle the same two special cases that are handled below for constant values:

1. `properties.key = NULL` expressions need to be rewritten to `not(has(property_group_column, 'key'))`, and
2. `properties.key = ''` expressions need to be rewritten to `has(property_group_column, 'key') and property_group_column['key'] = ''`
3. everything else can be evaluated as `property_group_column['key'] = value`

…but the analyzer doesn't seem to be able to figure out how to select the indexes for that query:

```
EXPLAIN indexes = 1
SELECT count()
FROM events
WHERE if(('xyz' AS value) IS NOT NULL, if(value = '', has(events.properties_group_custom, 'key' AS key) AND ((events.properties_group_custom[key]) = value), (events.properties_group_custom[key]) = value), NOT has(events.properties_group_custom, key))

Query id: 14aaa2b6-a8ff-45f5-aaf0-ee3159412f30

┌─explain────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))        │
│   Aggregating                                      │
│     Expression (Before GROUP BY)                   │
│       Filter (WHERE)                               │
│         ReadFromMergeTree (default.sharded_events) │
│         Indexes:                                   │
│           MinMax                                   │
│             Condition: true                        │
│             Parts: 5/5                             │
│             Granules: 9/9                          │
│           Partition                                │
│             Condition: true                        │
│             Parts: 5/5                             │
│             Granules: 9/9                          │
│           PrimaryKey                               │
│             Condition: true                        │
│             Parts: 5/5                             │
│             Granules: 9/9                          │
└────────────────────────────────────────────────────┘

18 rows in set. Elapsed: 0.003 sec.
```

... even though the there is only one branch that is used (`(events.properties_group_custom[key]) = value)`), and that condition when provided alone can use the index:

```
EXPLAIN indexes = 1
SELECT count()
FROM events
WHERE (events.properties_group_custom['key']) = 'xyz'

Query id: 53866aad-6bbc-41c3-8a24-071c688b446d

┌─explain─────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))         │
│   Aggregating                                       │
│     Expression (Before GROUP BY)                    │
│       Filter (WHERE)                                │
│         ReadFromMergeTree (default.sharded_events)  │
│         Indexes:                                    │
│           MinMax                                    │
│             Condition: true                         │
│             Parts: 5/5                              │
│             Granules: 9/9                           │
│           Partition                                 │
│             Condition: true                         │
│             Parts: 5/5                              │
│             Granules: 9/9                           │
│           PrimaryKey                                │
│             Condition: true                         │
│             Parts: 5/5                              │
│             Granules: 9/9                           │
│           Skip                                      │
│             Name: properties_group_custom_keys_bf   │
│             Description: bloom_filter GRANULARITY 1 │
│             Parts: 0/5                              │
│             Granules: 0/9                           │
│           Skip                                      │
│             Name: properties_group_custom_values_bf │
│             Description: bloom_filter GRANULARITY 1 │
│             Parts: 0/0                              │
│             Granules: 0/0                           │
└─────────────────────────────────────────────────────┘

28 rows in set. Elapsed: 0.003 sec. 
```